src/corecel/cont/Array.hh

```diff
@@ -45,88 +49,116 @@ struct Array
-    CELER_CONSTEXPR_FUNCTION const_pointer data() const { return data_; }
-    CELER_CONSTEXPR_FUNCTION pointer data() { return data_; }
+    CELER_CEF const_reference operator[](size_type i) const { return d_[i]; }
+    CELER_CEF reference operator[](size_type i) { return d_[i]; }
+    CELER_CEF const_reference front() const { return d_[0]; }
+    CELER_CEF reference front() { return d_[0]; }
+    CELER_CEF const_reference back() const { return d_[N - 1]; }
+    CELER_CEF reference back() { return d_[N - 1]; }
+    CELER_CEF const_pointer data() const { return d_; }
+    CELER_CEF pointer data() { return d_; }
+    //!@}
 
+    //!@{
     //! Access for structured unpacking
     template<std::size_t I>
-    CELER_CONSTEXPR_FUNCTION T& get()
+    CELER_CEF T& get()
     {
         static_assert(I < static_cast<std::size_t>(N));
-        return data_[I];
+        return d_[I];
     }
-
-    //! Access for structured unpacking
     template<std::size_t I>
-    CELER_CONSTEXPR_FUNCTION T const& get() const
+    CELER_CEF T const& get() const
     {
         static_assert(I < static_cast<std::size_t>(N));
-        return data_[I];
+        return d_[I];
     }
     //!@}
 
     //!@{
     //! \name Iterators
-    CELER_CONSTEXPR_FUNCTION iterator begin() { return data_; }
-    CELER_CONSTEXPR_FUNCTION iterator end() { return data_ + N; }
-    CELER_CONSTEXPR_FUNCTION const_iterator begin() const { return data_; }
-    CELER_CONSTEXPR_FUNCTION const_iterator end() const { return data_ + N; }
-    CELER_CONSTEXPR_FUNCTION const_iterator cbegin() const { return data_; }
-    CELER_CONSTEXPR_FUNCTION const_iterator cend() const { return data_ + N; }
+    CELER_CEF iterator begin() { return d_; }
+    CELER_CEF iterator end() { return d_ + N; }
+    CELER_CEF const_iterator begin() const { return d_; }
+    CELER_CEF const_iterator end() const { return d_ + N; }
+    CELER_CEF const_iterator cbegin() const { return d_; }
+    CELER_CEF const_iterator cend() const { return d_ + N; }
     //!@}
 
     //!@{
     //! \name Capacity
-    CELER_CONSTEXPR_FUNCTION bool empty() const { return N == 0; }
-    static CELER_CONSTEXPR_FUNCTION size_type size() { return N; }
+    CELER_CEF bool empty() const { return N == 0; }
+    static CELER_CEF size_type size() { return N; }
     //!@}
 
     //!@{
     //! \name  Operations
 
     //! Fill the array with a constant value
-    CELER_CONSTEXPR_FUNCTION void fill(const_reference value)
+    CELER_CEF void fill(const_reference value)
     {
         for (size_type i = 0; i != N; ++i)
-            data_[i] = value;
+        {
+            d_[i] = value;
+        }
     }
     //!@}
+
+  private:
+    T d_[N];  //!< Storage
 };
 
+//---------------------------------------------------------------------------//
+// DEDUCTION GUIDES
+//---------------------------------------------------------------------------//
+
+// Note: this differs from std::array, which deduces from T rather than the
```

Comment:
Probably I just lost in the English grammar ... does the 'which' applies/describes `std::array` (my reading) or `celeritas::Array` (what the text seems to describe).

Reply:
Your reading is correct: `std::array<T, N>`; vs `celeritas::Array<std::common_type_t<T, Us...>, N>`. Maybe you missed "rather than"?

Reply:
No really, what I actually missed is the existence of `std::common_type_t` for `celeritas::Array`.